lektor/admin/modules/livereload.py

```diff
@@ -0,0 +1,56 @@
+import json
+import secrets
+import threading
+from typing import Any, Generator, Optional
+
+from flask import Blueprint, current_app, request
+from werkzeug.exceptions import NotAcceptable
+
+PING_DELAY = 1.0
+RELOAD_DEBOUNCE_TIME = 0.05  # seconds
+
+bp = Blueprint("livereload", __name__, static_folder="static")
+# Use a random ID to detect reloading, which will be changed after reloaded
+version_id = secrets.token_urlsafe(16)
+
+reload_event = threading.Event()
```

Comment:
The `Event` wraps a `Condition` inside:

```python
    def wait(self, timeout=None):
        with self._cond:
            signaled = self._flag
            if not signaled:
                signaled = self._cond.wait(timeout)
            return signaled
```

When multiple SSE views are waiting for the same event(blocked at line 5), a flip of the event will trigger the reload message for all waiters. Unless it comes late to see the flag already cleared.

Reply:
> Unless it comes late to see the flag already cleared.

Well, yes.  That's my point: there's a race.

In this case, we don't really care about the value of the flag.  We just want to wake up all the threads that are waiting for a rebuild.  Using just a plain `Condition` will do exactly this.  (Use Condition.notify_all to wake up all the waiting threads.)

Reply:
> We just want to wake up all the threads that are waiting for a rebuild. Using just a plain Condition will do exactly this. (Use Condition.notify_all to wake up all the waiting threads.)

What I mean is, if the view is already in the waiting queue, `Event` also does the trick under the hood. But if it fails to catch the last bus, it will also miss the `Condition.notify_all`. Anyway t is true that `Event = flag + Condition`, we can ignore the flag and use a plain condition instead, but it seems not that different from the current approach, but I am Okay to change it both ways.

Reply:
Having thought about this overnight, you're right in that using an `Event` will work _most of the time_.  But neither `Event` nor a plain `Condition` is actually correct.

The listener threads (the thread(s) servicing the `/__reload__/events` request(s)) do external I/O. Specifically, they are responsible for sending data to the HTTP clients.  As such, they can spend significant periods of time waiting for I/O to complete.  If a rebuild happens while the thread is blocked for I/O (rather than waiting in the `Event.wait` call), then, if there are multiple listeners, it is likely to miss the notification, as one of the other listeners will likely see the event and clear the flag before the blocked thread gets a chance to `wait` on the event.  Using a lone `Condition` is even worse in this regard, since the event will definitely be missed by all threads that aren't blocked in the `Condition.wait` call.

A better solution would be to use a counter protected by a `Condition`.  On each build the counter gets incremented.  Every time a listener sees a counter value it hasn't yet seen, it knows a build has happened.

Reply:
SGTM, I will try this approach tomorrow.